### ISSUE TITLE
Feature Request - agent attributes - phase two

### ISSUE THREAD

**Issue by dustymc** [ArctosDB/arctos#7352]
**Is your feature request related to a problem? Please describe.**

Agent attributes exist and are running in production. Next step is to move everything else into the unified structure.

**Describe what you're trying to accomplish** 

Unity and happiness.

**Describe the solution you'd like**

Create attributes corresponding to each item in:

* https://arctos.database.museum/info/ctDocumentation.cfm?table=ctaddress_type
* https://arctos.database.museum/info/ctDocumentation.cfm?table=ctagent_name_type
* https://arctos.database.museum/info/ctDocumentation.cfm?table=ctagent_relationship
* https://arctos.database.museum/info/ctDocumentation.cfm?table=ctagent_status

plus

* agent.agent_remarks
* agent.curatorial_remarks
* agent.preferred_agent_name

Include appropriate controls and rules as necessary.

* make sure shipping addresses remain protected, whatever that looks like in this model
* filter ORCID, wikidata, etc.
* limit (or document?) the attribute correspondening to agent.preferred_agent_name to preferred name change history

Remove:

* agent.last_edit_by
* agent.last_edit_date

The new model will provide a more robust way to track these (along with WHAT was edited).

Drop:

* unique constraint on agent.preferred_agent_name, https://github.com/ArctosDB/arctos/issues/4534

I will adjust the UI to follow https://github.com/ArctosDB/arctos/issues/6813 and friends while I'm changing things. Primarily, that's no secondary data necessary for agent creation, and no requirements or presumptions of uniqueness on preferred name.

**Describe alternatives you've considered**

I could open issues for each individual term/action, but I think one issue for all is probably easier to follow.

**Additional context**

Specific concerns (eg stuff mentioned in https://github.com/ArctosDB/arctos/issues/7318) can be addressed as necessary during or after this migration.

**Related/Included**

* https://github.com/ArctosDB/arctos/issues/6699 - will just work in new model
* https://github.com/ArctosDB/arctos/issues/7322 - will be released with new model, necessary in an environment with nonunique agents.

**Priority**

Critical - we are now in an intermediate model with denormalized data and an overly-complex UI, and important data are almost certainly still being stuffed into remarks.

@mkoo @Jegelewicz someone anyone please say GO!

**Comment by mkoo** [comment]
This is a great consolidation of all the various issues and a nice roadmap to get to our new goal of Agents.
Please start on the tasks in its logical order as it has been approved in AWG and Agent Committee meetings (among other committee meetings)

The only one I dont see in the chain of requirements (create list) is this one (unless it is implicit):
https://github.com/ArctosDB/arctos/issues/6813#issuecomment-1899522184
which allows the check on verification/ minimum requirements for agents. This needs to be included

**Comment by dustymc** [comment]
>verification

Sorry, things got a little convoluted. https://github.com/ArctosDB/arctos/issues/6813#issuecomment-1899522184 became https://github.com/ArctosDB/arctos/issues/7318 which is in production now (but not DOING much as it's behind this issue). If there's something about the UI that could be better (other than all the legacy stuff piled on top, of course), please let me know NOW - it'll be the only UI once this is done.

I of course gave myself Attribute One and a gold star.

<img width="985" alt="Screenshot 2024-01-31 at 11 52 55" src="https://github.com/ArctosDB/arctos/assets/5720791/aac53700-4711-4a66-96c3-535e632d7fdd">

I'll run this through the dev channel for more exposure and get started.

**Comment by mkoo** [comment]
Ah cool! Testable in test then??? Can we have a demo of the progress of these tasks at the Issues meeting on Thursday?

As for #6813/ #7318 ideally this is a largely automated process based on specific criteria which I dont know if it has been on the agenda of the agents or AWG committee, but probably should be.

**Comment by Jegelewicz** [comment]
> Testable in test then???

It's in production

**Comment by dustymc** [comment]
>Testable

Yep, in prod.

>demo

Doesn't do much yet, just the model and UI made it through the Phase One process. But yea, if someone wants to complain about the UI (or start planning for a replacement - I think @Jegelewicz has ideas) or point out something that the model could but does not do or etc now's the time.

MAYBE I'll have something more in test by then, but probably going to be all behind-the-scened for a few days. I'm working on https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctagent_attribute_type now.

> largely automated process

No problem, just file an Issue with the criteria (after this) and I'll set up a bot, or periodically script things in, or whatever turns out to be best.

**Comment by Jegelewicz** [comment]
> if someone wants to complain about the UI (or start planning for a replacement - I think @Jegelewicz has ideas)

I hadn't thought about it, but since I can see what you have. Let me propose something.

Can we make the agent search have the same look and feel as the catalog record search? Search form at the top, results at the bottom - this could let you see and possibly compare results to make better choices. So each row of the results would have something like this:

preferred name | given names | family names | born | died | ORCiD | Wikidata | Library of Congress | employment 
 -- | -- | -- | -- | -- | -- | -- | -- | -- 
[Teresa J Mayfield-Meyer](https://arctos.database.museum/agents.cfm?agent_id=21300608) | Teresa | Jegelewicz; Mayfield-Meyer | 1800-01-01 | | https://orcid.org/0000-0002-1970-7044 | | | Collection Manager at [University of Texas - El Paso](https://arctos.database.museum/agents.cfm?agent_id=21277509); Community Coordinator at [Arctos Consortium](https://arctos.database.museum/agents.cfm?agent_id=21334341) 

Sorting by column headers would allow for easier comparisons as would the ability to check one or more and only see those (remove some rows like you can in catalog record search).

Editing would be a separate page, just like a catalog record. So from the results, if I wanted to edit my Agent record, I would click my name and end up at the edit page (or the public profile if I am a public user).

This means we should also think about how or if the search should be different for public users and editors. I don't think it needs to be? Just hide the appropriate attributes from anyone without manage_agents?

Next up - organizing the edit page.

**Comment by dustymc** [comment]
A full-width page would certainly work better for editing.

They data aren't very tabular, I don't think your example is realistic, but I can figure some sort of less-bare results out (and it'll be easy to modify as we find things that could be better).

>remove some rows

Seems like wild overkill (that sort of thing is hugely complex), but see above - if we need to we can.

>organizing the edit page


<img width="317" alt="Screenshot 2024-02-02 at 07 58 45" src="https://github.com/ArctosDB/arctos/assets/5720791/da33e52f-0e0c-4c6d-bc43-3941fde1f1c0">

**Comment by Jegelewicz** [comment]
This is the current demo

![image](https://github.com/ArctosDB/arctos/assets/5725767/df7e3db6-69c3-402a-8a9a-65740fbf58a3)

Once all of my information is in that format, it will just be a long list of hard to process double rows - however, if we weren't taking up half of the edit page with a blank white space because the search form is over there somewhere at the top, we might be able to fit each attribute on a single row - I think that would be a good start to making this less hard to deal with.

Still a long list of stuff without any breaks is hard on the eyes and the mind. If possible, it would be nice to organize the entered attributes by the broad categories - this would be similar to the current edit page and one familiar thing would be good.

I also think that the section for adding new information should be at the top of the edit form. If it starts with three rows (and has an add more) it won't be too intrusive and if I am coming to add a new attribute, I don't want to scroll to the bottom every time.

# Manage Agent

![image](https://github.com/ArctosDB/arctos/assets/5725767/b00f8ce6-dece-446c-93e8-5d535e49e989)

## Add Attributes

attribute define | attribute value | begin date | end date | related_agent | determined date | determiner | method | remark
-- | -- | -- | -- | -- | -- | -- | -- | --
 |  |  |  |  |  |  |  | 
add row

save

## Edit Attributes

save

### Names

attribute define | attribute value | begin date | end date | related_agent | determined date | determiner | method | remark
-- | -- | -- | -- | -- | -- | -- | -- | -- 
given name | Teresa |  |  |  | 2024-02-02 | jegelewicz | personal knowledge | 
family name | Jegelewicz |  |  |  | 2024-02-02 | jegelewicz | personal knowledge | 
family name | Mayfield-Meyer |  |  |  | 2024-02-02 | jegelewicz | personal knowledge | 

### Addresses

attribute define | attribute value | begin date | end date | related_agent | determined date | determiner | method | remark
-- | -- | -- | -- | -- | -- | -- | -- | -- 
ORCiD | https://orcid.org/0000-0002-1970-7044 |  |  |  | 2024-02-02 | jegelewicz | personal knowledge | 

### Milestones

attribute define | attribute value | begin date | end date | related_agent | determined date | determiner | method | remark
-- | -- | -- | -- | -- | -- | -- | -- | -- 
employment | Collection Manager |  |  | [University of Texas - El Paso](https://arctos.database.museum/agents.cfm?agent_id=21277509) | 2024-02-02 | jegelewicz | personal knowledge | 
employment | Community Coordinator |  |  | [Arctos Consortium](https://arctos.database.museum/agents.cfm?agent_id=21334341) | 2024-02-02 | jegelewicz | personal knowledge | 

### Statuses
attribute define | attribute value | begin date | end date | related_agent | determined date | determiner | method | remark
-- | -- | -- | -- | -- | -- | -- | -- | -- 
status | verified | 2024-02-01 |  |  | 2024-02-02 | jegelewicz | personal knowledge | 

save edits

**Comment by dustymc** [comment]
>double rows

That's up to your browser.

>broad categories 

That's why I just added that to the code table....

>adding new information should be at the top

Always has been??

**Comment by mkoo** [comment]
Not sure if you have another issue open for the test version of the agents manage/ edit/ public interface but some general comments. Overall I think it works but test was testy so havent really been able to try out everything I wanted to.

One thing I think needs to be changed on the public page is to have the remarks right under the top line of the name. That is where we have been adding all our biographical info (and its been a lot since we are importing our finding aid bios) so should be up top with the name and not midway or bottom half of page (save that area for the nitty gritty)

I'll return to this when test is more cooperative!

**Comment by dustymc** [comment]
>another issue

Nope, here's good, and thanks!

>the remarks

Details, probably, but: There is no THE, there can be 987 "remarks fields" (or anything else) in this model, and they can all carry (maybe critical) metadata - nothing really survives this in 'just a string' form. Also....

>adding all our biographical info

... a place dedicated to that is super-trivial (just a code table request), maybe we should do that, move some stuff, and make that (rather than the 'junkdrawer data' remarks is meant to hold) the prominent thing? That'd have to be post-release, happy to help shuffle, @AJLinn might want in on that conversation.

> test is more cooperative!

It should be happy now. It is slow, but I think that's just test - I tried bits individually at prod and they're fast there. It'll also be different based on who you are - eg https://arctos-test.tacc.utexas.edu/agent/280 will be **the** agent detail page (possible because of how I structured https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctagent_attribute_type). Similarly https://arctos-test.tacc.utexas.edu/agent.cfm will be **the** search page, users with different rights will just be able to access different things.

**Comment by Jegelewicz** [comment]
So search will now be unified? This seems odd

![image](https://github.com/ArctosDB/arctos/assets/5725767/f55b0e09-2bc1-4730-b73a-dacea3cb4125)

I would expect my agent to show up in results

**Comment by Jegelewicz** [comment]
But the "new" isn't using old data yet?

This looks good

![image](https://github.com/ArctosDB/arctos/assets/5725767/b37943ed-de8c-41a6-b90d-f05fb31e4dcf)

**Comment by Jegelewicz** [comment]
OK so the search name thing is using "starts with"

![image](https://github.com/ArctosDB/arctos/assets/5725767/7fc95d04-38d0-4eb3-b7ec-d7ca654452bc)

I think the old search was "any part of any name". I guess everyone would get used to it, but that single change might make a lot of people upset?

**Comment by Jegelewicz** [comment]
I would remove the admonishing frowny face here. I don't want people knowing my birthdate - after I'm dead you guys can add it, but for now I'd like to keep that to myself if possible....don't make me feel bad about that.

![image](https://github.com/ArctosDB/arctos/assets/5725767/73101c48-cdc3-422e-9bf6-57b892571249)

**Comment by Jegelewicz** [comment]
Buttons at the top that let me scroll to one of the sections? Like at the top of the catalog record page.

![image](https://github.com/ArctosDB/arctos/assets/5725767/ec2a91f5-b6c8-4901-977a-25f4894b3115)

also here

![image](https://github.com/ArctosDB/arctos/assets/5725767/d408113a-d508-4dec-b202-17b0abad230f)

**Comment by dustymc** [comment]
>search will now be unified? 

That's the plan - which means some terms won't work for some users. (Or there's a bug, I'll investigate after I figure out why edit has gone crazy....)

>"new" isn't using old data yet?

hu?

> is using "starts with"

No, but it is using...

```
 <cfif len(agent_name) gt 2>
	                    and agent.preferred_agent_name ilike <cfqueryparam value="%#agent_name#%" cfsqltype="cf_sql_varchar">
	                </cfif>
```

I'll do something or another....

>keep that to myself 

That's fine, but not having some sort of information will also make it a bit more likely that someone will get confused and add another you (at which point you cannot get proper attribution for your work/donations/whatever). 

>at the top

Everybody wants everything at the top.... - IDK, easy to add, but not going to be my focus for a while (and see just above - a bunch of them will probably scroll to nowhere for a bunch of agents, unless you're asking for a bunch of prefetch and everything to be different everywhere...)

**Comment by Jegelewicz** [comment]
email addresses and phone numbers (except work phone) should not appear on the public profile? Also, let's make sure that the "attributes details" doesn't reveal personal information when viewed by those without manage_agents.

**Comment by Jegelewicz** [comment]
> will probably scroll to nowhere

Why - the header is there even when there is no data (as per my frowny comment) - just scroll to the header?

**Comment by dustymc** [comment]
>should not appear on the public profile?

That's just data, check the code table (and let me know if I should do something different for the migration)

>viewed by those without manage_agents

Also data, you have admin directory (but I still need to know if there's a migration-leak)

>frowny comment

You're mixing up two pages....

**Comment by Jegelewicz** [comment]
> You're mixing up two pages....

OK - but when a section is not there in the catalog record the button just isn't there - right? Same principle?

**Comment by Jegelewicz** [comment]
> check the code table

Yes - so perhaps for these code tables we need to be able to mark these as public or not? 

[ctaddress_type](https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctaddress_type)
[ctagent_name_type](https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctagent_name_type)
[ctagent_status](https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctagent_status)
[ctagent_relationship](https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctagent_relationship)

and
[ctagent_rank](https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctagent_rank) should all be limited to Arctos operators? or only those with manage_agents?

better ideas?

**Comment by Jegelewicz** [comment]
> It's clutter, it's handy to know agent has no whatevers

Yeah - I see. If it were me, I would say, let's do whatever we are doing in catalog records for consistency, but others might have different ideas.

**Comment by Jegelewicz** [comment]
> None of those tables survive this.

Thanks - Now I see https://arctos-test.tacc.utexas.edu/info/ctDocumentation.cfm?table=ctagent_attribute_type and the public thing is there. We should start with whatever was not public under the old system (however that was accomplished) and make sure those things are not public in the new. Is there any easy way for me to know that?

**Comment by Jegelewicz** [comment]
Now let's talk about the search form

![image](https://github.com/ArctosDB/arctos/assets/5725767/0bce7721-28cf-4fc7-ace4-f57e02e41c49)

Can we not have the fields in a long list that requires scrolling?

How about

# Find Agents

Agent Name (three characters required to search) | Include verbatim agent?
-- | -- 
![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0)
 | ![image](https://github.com/ArctosDB/arctos/assets/5725767/5687658d-3087-4e37-8f7c-441e6244d3f2)

Attribute | Attribute Value (three characters required to search, prefix with = for exact) | Begin date (default on-or-after, prefix with = for exact) | End date (default on-or-before, prefix with = for exact) | Method | Remark
 -- | -- | -- | -- | -- | -- 
![image](https://github.com/ArctosDB/arctos/assets/5725767/50c272aa-9c4f-4010-8de7-04b779afa090) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) 

Determined Date (default startswith, prefix with = for exact) | Determiner Agent | Related Agent | Creator Agent
-- | -- | -- | -- 
![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) | ![image](https://github.com/ArctosDB/arctos/assets/5725767/4fff97cd-4e51-4e38-b95d-7924f63456b0) 


Just make it look like 

![image](https://github.com/ArctosDB/arctos/assets/5725767/9e2be489-8a76-4e31-8b28-2455b451021b)

**Comment by dustymc** [comment]
Addresses created with old macs are now (I so hope...) being handled properly and not making lots of not-quite-duplicates.

I moved remarks (for @mkoo) and identifiers (for me) up. We should still move biography out of remarks.


<img width="1729" alt="Screenshot 2024-02-07 at 09 24 01" src="https://github.com/ArctosDB/arctos/assets/5720791/b53205b9-ace1-4429-b164-4d8f28b9c7f8">

There are lots of linefeeds and such in addresses, not sure I see much reason to change that, we can allow them in remarks and then 'htmlify' for display and...


<img width="1729" alt="Screenshot 2024-02-07 at 09 39 30" src="https://github.com/ArctosDB/arctos/assets/5720791/5d32f02b-58bc-44ec-beab-9b44169cc1a0">

I'll open an issue, this is a mess but it doesn't have to be resolved before production.

**Comment by dustymc** [comment]
> talk about the search form

It's different.

* There's a search anything (ish) option
* There are indexes to support that
* There's some flexy css which a designer could work with

<img width="589" alt="Screenshot 2024-02-07 at 11 13 21" src="https://github.com/ArctosDB/arctos/assets/5720791/b576751c-4c6a-4e80-8a17-35eb54657524">

and my [styling](https://github.com/ArctosDB/arctos/issues/7352#issuecomment-1932565037) (see also https://github.com/ArctosDB/arctos/issues/7383) doesn't play very nice with some data....

<img width="171" alt="Screenshot 2024-02-07 at 11 14 24" src="https://github.com/ArctosDB/arctos/assets/5720791/d1289256-3a75-404e-9017-6b368d300197">

Bio/bio reminds me of [Chile](https://en.wikipedia.org/wiki/Biob%C3%ADo_Region)....

**Comment by Jegelewicz** [comment]
A few tweaks

Can the verification icons appear here?

![image](https://github.com/ArctosDB/arctos/assets/5725767/11ec73a9-257c-4f29-8ade-accede712dbd)

repeat the save button at the top

![image](https://github.com/ArctosDB/arctos/assets/5725767/8f8324aa-5e40-4d2b-af8f-2f9745070df4)

if all I am doing is adding something, the scroll to the bottom is a time sink.

Can the address section
![image](https://github.com/ArctosDB/arctos/assets/5725767/039947d6-d593-4790-aa48-7d77f2401ebc)

be like this instead? I think it would be more readable.

correspondence: blah blah blah
correspondence: blah blah blah blahblah
work phone: xxx-xxx-xxxx

Things with end dates should be greyed out?

I like this - I hope others do too!!!

**Comment by dustymc** [comment]
> like this instead? 

See also https://github.com/ArctosDB/arctos/issues/7383

**Comment by Jegelewicz** [comment]
Would it break anything to put a couple of `<div>` in here to make this a little more readable?

![image](https://github.com/ArctosDB/arctos/assets/5725767/ac786215-521c-428c-9153-ebca63afdc43)

basically one Div with

Search Most Anything
Agent Name
Include verbatim agent?
Include Duplicates 
Agent Creator
Agent Create Date

and another with

Attribute
Attribute Value
Begin date
End date
Determined Date
Attribute Determiner
Related Agent
Method
Remark

This keeps the general stuff up top and the attribute stuff below. 

Some terms including "Attribute" and some not seems odd? The one I think might need it more is Remark, because I can see that being confused with the old Agent Remark.... What if the `<div>` had a heading and we did this instead?

## Agent Attributes
Category
Value
Begin date
End date
Determined Date
Determiner
Related Agent
Method
Remark

Just thinking about user experience. Not saying my way is perfect, but maybe easier for people to navigate?

![image](https://github.com/ArctosDB/arctos/assets/5725767/8bc86173-eef3-40fe-b3bf-680b61341a3b)

**Comment by dustymc** [comment]
The search form is set up so such things are pretty easy to shuffle, and hopefully pretty hard to break.

I'm not convinced that's the right split - I suspect the use will be 'anything' about 99% of the time and 'other stuff' very occasionally. (So maybe an accordion or similar - which would be nice if we were using bootstrap, but then we'd be stuck with bootstrap...)

You've got at least two attribute-things in the not-attributes section, which doesn't seem ideal. 

The breaks add unnecessary whitespace, and 50.00% of the time I hear that's evil. (Everyone else has giant screens and wants more whitespace.)

IDK either, seems like I can spend all eternity shuffling things around and still end up with about the same number of people (but different people, I hope!) all up in arms. Where's that UI designer we've been talking about forever?!

Your other suggestions are (mostly) implemented. (And there's another 50/50 split: Some will want the summary - a grayed-out email address is all they need - and others will want a table-or-something with all the gritty details. Things are kinda leaning towards the former at the moment.)

**Comment by Jegelewicz** [comment]
> You've got at least two attribute-things in the not-attributes section, which doesn't seem ideal.

Ah - ha I was fooled by the "Agent" part of those two things into believing they meant who created the AGENT and when. Shouldn't those really be Attribute Creator and Attribute Created Date?

**Comment by ebraker** [comment]
Just checked out the demo. Unified Agent Search - Cool!

Question on the edit page. Will the entered data sort into obvious groupings? e.g. Birth and death date be displayed together, all of the name variants grouped together, relationships grouped together, addresses grouped together, or will they just appear in the order they are entered? I think it is important to have things sorted so that it is easy to navigate. Right now everything is well-organized in obvious blocks and it makes the page easy to work with and corroborate agent data quickly because I can know where to "jump to" to look for the info I need.

**Comment by Jegelewicz** [comment]
@ebraker look at my profile in test and you can see how everything sorts into blocks. (of course, you'll have to log in to test and hit edit).

https://arctos.database.museum/agent/21300608

![image](https://github.com/ArctosDB/arctos/assets/5725767/1d836717-a836-459b-a2a3-83f90020b39a)

**Comment by dustymc** [comment]
>error

Should be fixed, I'm still tuning some stuff up, hopefully the interruptions will be few and localized from here....

> name get pushed

yep

>what would happen if I enter edits?

You can probably confuse yourself! I could force-overwrite with preferred name or something, but in general there's definitely less control on my end. I think that's probably a better reflection or reality, but it's also an opportunity to make huge messes. I'm *really* hoping that The Committee will develop some documentation-or-whatever, maybe watch for not-so-great data, etc. 

>visual clutter

See above and in related issues (eg https://github.com/ArctosDB/arctos/issues/7359#issuecomment-1921857683) - I'm hoping we end up with a lot fewer, more 'mergey' attributes, and of course I hope they're all accompanied by good metadata. I'd be very resistant to limiting the possibilities until that's more settled. That said, this is all running through a full-authentication API - should be pretty trivial for anyone to throw up any sort of UI.

**Comment by AJLinn** [comment]
I agree with @ebraker - I'm having a really hard time focusing on anything because of the low contrast between lines and various categories.

Is it possible to  have the column headers float down the page as you scroll down? That might not be in our capability, but at least having it repeated in each section grouping would be helpful since once you get to the bottom of a well-documented agent, you have to keep scrolling  up to see what you're supposed to put in the field. Once you data in a field with a text hint, you can no longer see the hint.

I'm curious about the overall responsiveness of the page. I'm on a fairly widescreened iMac but it seems like it's taking all the real estate. If I wanted to have one window open with this and one window with my catalog page or data entry page, I'd have a hard time seeing a lot. I also wonder about viewing it on a smaller laptop screen.

**Comment by dustymc** [comment]
>low contrast

I think that's probably a subject for https://github.com/ArctosDB/arctos/issues/4874, I'm using a "standard" color ``var(--arctoslightblue)`` that came from UI discussions, I don't think breaking what little standardization we have is a great idea!

>column headers float 

Not really, that would need some other UI.

>hint

If you mean the placeholders, I could replicate them in a title (so you'd see them on hover)?

>smaller laptop

All of Arctos is written on a little laptop.

>repeated in each section grouping 

Definitely possible, would cost some vertical real estate, maybe that's not important for most agents??

**Comment by AJLinn** [comment]
>   repeated in each section grouping

Definitely possible, would cost some vertical real estate, maybe that's not important for most agents??

I think I'd prefer a longer page if I didn't have to keep scrolling to the top to see what information I'm looking at in each section. Open to other opinions on the options tho!

**Comment by Jegelewicz** [comment]
> I'm using a "standard" color var(--arctoslightblue) that came from UI discussions, I don't think breaking what little standardization we have is a great idea!

Instead of alternating it with white, alternate with arctosdarkblue as (I think) on the catalog record results.

**Comment by dustymc** [comment]
It's stripey-ier with repeating "headers."

**Comment by Jegelewicz** [comment]
Hmmm - not sure there is enough contrast, but I'll let other people look. I do like the repeating "headers"!

**Comment by Jegelewicz** [comment]
Maybe there are too many lines

![image](https://github.com/ArctosDB/arctos/assets/5725767/fe68c326-14c2-4935-8a38-51a5a28d4d25)

vs

![image](https://github.com/ArctosDB/arctos/assets/5725767/cf7e5d64-9423-4303-bb54-34d159a8ed1c)

Not saying you should change anything NOW - just offering up an observation for others to consider.

**Comment by dustymc** [comment]
>too many lines

If you mean the adds: I can easily change that, just give me a number.

If you mean the line height: It varies with data.

<img width="583" alt="Screenshot 2024-02-19 at 07 30 56" src="https://github.com/ArctosDB/arctos/assets/5720791/9fbb15b4-6475-423b-81e8-3d4c1a69a353">

(And you're comparing a native edit UI with a display-centric library).

**Comment by Jegelewicz** [comment]
I mean the double lines between everything - that seems cluttering and breaks up the row color too much to make it useful for me.

**Comment by dustymc** [comment]
>double lines

Ah, thanks - it's easy enough to override the native 'table border' with css....


<img width="544" alt="Screenshot 2024-02-19 at 07 45 55" src="https://github.com/ArctosDB/arctos/assets/5720791/2a1d766a-9b47-4e62-8b9e-ee750a35760f">

but then we'd have ONE table like that (and it's probably not that simple at scale, seems like it never is), and about a bajillion not like that. Seems like a [Phase Three](https://github.com/ArctosDB/arctos/issues/4874#issuecomment-1199331785) problem at the moment....

**Comment by ebraker** [comment]
I think now that headers are repeated we could do without the greyscale prompts within the empty fields to make things less busy.

![image](https://github.com/ArctosDB/arctos/assets/11336485/ad93d270-27a2-4cdb-a1af-3aebd3f63e4c)

**Comment by ebraker** [comment]
> See above and in related issues (eg https://github.com/ArctosDB/arctos/issues/7359#issuecomment-1921857683) - I'm hoping we end up with a lot fewer, more 'mergey' attributes, and of course I hope they're all accompanied by good metadata. I'd be very resistant to limiting the possibilities until that's more settled. That said, this is all running through a full-authentication API - should be pretty trivial for anyone to throw up any sort of UI.

Yes but I think we need to fix this on our end vs have people create their own UIs (that is beyond most people's capabilities, certainly mine). Right now I don't think we should roll out these changes before figuring out these UI issues as the page is much more difficult to navigate than current Manage Agent page, which sort of nullifies any gains in functionality.

**Comment by ebraker** [comment]
Ok, just tested and there will certainly be issues for relationship agents as I was able to enter and save free text in the value field.

![image](https://github.com/ArctosDB/arctos/assets/11336485/160160ae-701a-4d79-b1a2-f47b5fc8a6b8)

Also, is it possible to make the related agents blue hypertext vs the arrows? I think it makes the linkages more apparent and that the names "do stuff."

![image](https://github.com/ArctosDB/arctos/assets/11336485/5b2ba1ec-ef32-43d2-a47b-d78383d41799)

**Comment by ebraker** [comment]
> It's not broken, it's https://github.com/ArctosDB/arctos/issues/5172 - this addresses a stated need for more functionality, it's not just different UI, and it's not the same capabilities as the old UI. "There, doing that, with them, then" was not possible before and is now. That requires a different UI.

I'm not saying it's broken. Simply that big changes can frustrate people so we need to have some sort of basic visual decluttering of this page. Yes we are gaining functionality, but if no one uses it because it is too intimidating/overwhelming, then we have lost something. I don't think it would take that much to make the form a little less unwieldy.

**Comment by DellaCHall** [comment]
Overall, I am a fan of this change - thank you! I have a few questions/suggestions for improvements:

**1) Agent auto-populate issue:** I used to be able to type a name exactly and it would just populate into any agent/determiner field, if there were no duplicates. Now it seems to search every field, including relationships, so all people/organizations associated with an agent also show up, then I have to click the correct one. This added extra step seems silly when typing a full name exactly, and might actually deter people from adding relationships, due to the annoyance (for example, I'll probably remove my relationship with my husband, so I can enter data faster). Is there a way to have it auto-populate the agent if there are no duplicates in the name fields, like it used to? I'm not sure what is gained by this particular change, so maybe I just need a different perspective to understand it.

**2) Search Results Table:** Could the search result table headers be customizable like the record search page? I'm not sure why those particular fields are the default headers, but they're not what I would choose. Also, could this table of data be exportable? Could it be sortable? Especially by name types? I have always disliked that the data sorts automatically by first name. As others have mentioned, right now it's just a really long list to scroll through...not very usable.

**3) Remarks field:** I see there was already some discussion with @mkoo about biographies that have been previously stored in the remarks field. Hoping there will be an easy way for me to migrate that data to the new biographies field (which I like!). Remarks currently show up in the search results table under "Other", but not on an Agent's individual summary page. Can Remarks display on the agent summary page, too? Previously it displayed.

**4) Verbatim agent:** I think the search question of "Include verbatim agent?" should default to yes instead of no.

**5) Hopeful question:** Is this a first step in reintegrating verbatim agents at some point down the line?

**Comment by ewommack** [comment]
> Agent auto-populate issue: I used to be able to type a name exactly and it would just populate into any agent/determiner field, if there were no duplicates. Now it seems to search every field, including relationships, so all people/organizations associated with an agent also show up, then I have to click the correct one. This added extra step seems silly when typing a full name exactly, and might actually deter people from adding relationships, due to the annoyance (for example, I'll probably remove my relationship with my husband, so I can enter data faster). Is there a way to have it auto-populate the agent if there are no duplicates in the name fields, like it used to? I'm not sure what is gained by this particular change, so maybe I just need a different perspective to understand it.

Try entering your user name instead of your full name. That works for me to only pull up me. With the relaxation of allowing people to have the same name now we're going to hit into having two or even more of the same person coming up in an exact name search as this as we get more and more people with the same or very similar names brought into the system. I think it is probably safer to have us search for people unless you know exactly who you want.

**Comment by Jegelewicz** [comment]
> Is this a first step in reintegrating verbatim agents at some point down the line?

@DellaCHall yes it is.

**Comment by Jegelewicz** [comment]
> I'll probably remove my relationship with my husband, so I can enter data faster

NOOOOOOOO! Do what Beth says and use your Arctos username.

**Comment by Jegelewicz** [comment]
> why those particular fields are the default headers

They are the groups of attributes as they currently exist in the edit form (mostly)

> they're not what I would choose.

What would you choose?

> could this table of data be exportable? 

Probably - @dustymc could say

> Could it be sortable? Especially by name types?

Name types are not one of the columns, so maybe this is something that would come from my question above. But now, people could have more than one "first name" attribute, so this may not be easy to handle.

**Comment by Jegelewicz** [comment]
> I think the search question of "Include verbatim agent?" should default to yes instead of no.

This is probably one of those 50/50 decisions - half will want it default yes and half no and we have no way to please everyone.

**Comment by Jegelewicz** [comment]
> Used by Collection: This field was previously an option on Manage Agents as a search field, and it appears to be missing from this new iteration. I liked that field, can it be added to this?

@dustymc possible?

**Comment by DellaCHall** [comment]
> Try entering your user name

I appreciate the suggestion of using my Arctos username! I can certainly do that for myself. And I won't remove my relationship to my husband, @Jegelewicz  :)

> I think it is probably safer to have us search for people unless you know exactly who you want.

Sure, but when I do know exactly who I want, I think I should be able to type their name and get just them (so long as there are no duplicates of the name fields). This is a super basic function many databases have (and, Arctos currently has!).

> having two or even more of the same person coming up in an exact name search as this as we get more and more people with the same or very similar names brought into the system

I understand. But these aren't in the name field of the agent - they're not even close to the same names. As an example, the agent Alaska Department of Fish and Game has a lot of "associate of" relationships. One of those is someone named "Doreen I. Parker McNeill". If I type in "Doreen I. Parker McNeill" as a determiner on an agent record attribute, why should it pull up a big list of ADF&G associates I have to scroll past, just to get to what I typed in to click on it? I stand by my suggestion that agent fields should auto-populate (as they do now) if there are no duplicates in the **name fields**, specifically.

**Comment by Jegelewicz** [comment]
> these aren't in the name field of the agent - they're not even close to the same names. As an example, the agent Alaska Department of Fish and Game has a lot of "associate of" relationships. One of those is someone named "Doreen I. Parker McNeill". If I type in "Doreen I. Parker McNeill" as a determiner on an agent record attribute, why should it pull up a big list of ADF&G associates I have to scroll past, just to get to what I typed in to click on it? I stand by my suggestion that agent fields should auto-populate (as they do now) if there are no duplicates in the name fields, specifically.

Can we just eliminate related agents from the pick searches? I think that would do it.

**Comment by DellaCHall** [comment]
> They are the groups of attributes as they currently exist in the edit form (mostly)

Okay, I see now. I was confused because I didn't see the new Bio field included...But I'm now seeing Profile, grouped under Other.

> What would you choose?

It would depend on the search I was trying to do and the specific data I needed, hence my request to customize/export/sort. That said, something useful for me would be splitting the names out first, middle, last (particularly for sorting, if that could be an option), and all the AKAs could group together. I'm not sure if everyone else would want that, but it would be extremely useful for me.

**Comment by dustymc** [comment]
I think there are at least two new Issues in here.

1. The pick can do whatever, there are functional implications to those things. (The unified format makes deep digs easier, I think I probably did whatever's there as an effort to reduce duplicate creation. I think that's deserving of dedicated discussion, clearly there's some difference in viewpoints and that's a globally-shared form.)
2. The old search returned a list of names - new does that (and some more), I think the details would benefit from more focused discussion than we're likely to achieve down here.

**Comment by dustymc** [comment]
>Agent auto-populate issue

https://github.com/ArctosDB/arctos/issues/7434

> Search Results Table

New Issue, I don't understand enough of your comments to kick that off. At this time it's just a simple table without much room to customize (but it's far from a list and never sorted by first name so maybe we're not talking about the same things??). I think @mkoo has maybe-not-immediate API-involved plans, that should be an opportunity to do WHATEVER.

>download

https://github.com/ArctosDB/arctos/issues/6890


>Hoping there will be an easy way for me to migrate that data to the new biographies field

There's CSV in https://github.com/ArctosDB/arctos/issues/7418, I'll update at release, PLEASE join in.

>Verbatim agent: I think the search question of "Include verbatim agent?" should default to yes instead of no.

https://github.com/ArctosDB/arctos/issues/5157

>reintegrating verbatim agents

Yes, this will allow low-information agents - but I still hope @ArctosDB/agents-committee will provide some guidelines, and this certainly won't change my recommendations regarding verbatim agents. Lots of discussion in https://github.com/ArctosDB/arctos/issues/7135.

>Used by Collection

done

**Comment by dustymc** [comment]
Stuff from https://docs.google.com/document/d/1lXFhBd4LH4802pOEGkj892ZUcJt19kh8Os1ky-qG570 implemented

**Comment by dustymc** [comment]
> bars would be dark blue

Crayons please, you've lost me. (Or we've come full circle and you're asking me to put back the thing that I was asked to remove a couple days ago...)

**Comment by Jegelewicz** [comment]
> crayons please

![image](https://github.com/ArctosDB/arctos/assets/5725767/319c7790-f6c0-49ff-b133-5d4efce7b785)

**Comment by dustymc** [comment]
oh my...
<img width="1668" alt="Screenshot 2024-02-20 at 08 55 32" src="https://github.com/ArctosDB/arctos/assets/5720791/6c1a05cc-2609-4eca-94d2-f2506733ec36">

**Comment by Jegelewicz** [comment]
That's OK - it makes it pretty clear that this agent has very limited data....

**Comment by ewommack** [comment]
The DEI Committee met today and reviewed the new Agent and Agent Attribute system. Overall we think that the new systems will be a step forward to opening up Agents to be more culturally inclusive.
We have four suggestions:
https://github.com/ArctosDB/arctos/issues/7437
https://github.com/ArctosDB/arctos/issues/7439
https://github.com/ArctosDB/arctos/issues/7133
and 
https://github.com/ArctosDB/arctos/issues/7440

**Comment by dustymc** [comment]
I nuked a bunch of full duplicates and made a sorta-unique index:

```
create unique index iu_compound_agent_attribute on agent_attribute (
    agent_id,attribute_type,md5(attribute_value),begin_date,end_date,related_agent_id
) nulls not distinct where deprecation_type is null ;
```
